cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,428 @@
+            has_inertia_node = true;
+            inertial_desc_.inertia_ = parseInertiaMatrixNode(node);
+        }
+    }
+
+    ASSERT(has_mass_node);
+    ASSERT(has_inertia_node);
+
+    return inertial_desc_;
+}
+
+UrdfVisualDesc UrdfParser::parseVisualNode(FXmlNode* visual_node)
+{
+    UrdfVisualDesc visual_desc_;
+
+    visual_desc_.name_ = TCHAR_TO_UTF8(*visual_node->GetAttribute(TEXT("name")));
+
+    bool has_origin_node = false;
+    bool has_geometry_node = false;
+    bool has_material_node = false;
+
+    for (auto& node : visual_node->GetChildrenNodes()) {
+        FString tag = node->GetTag();
+        if (tag.Equals(TEXT("origin"))) {
+            ASSERT(!has_origin_node);
+            has_origin_node = true;
+            visual_desc_.origin_.SetLocation(parseVector(node->GetAttribute("xyz")));
+            visual_desc_.origin_.SetRotation(parseRotation(node->GetAttribute("rpy")));
+        } else if (tag.Equals(TEXT("geometry"))) {
+            ASSERT(!has_geometry_node);
+            has_geometry_node = true;
+            visual_desc_.geometry_ = parseGeometryNode(node);
+        } else if (tag.Equals(TEXT("material"))) {
+            ASSERT(!has_material_node);
+            has_material_node = true;
+            visual_desc_.material_desc_ = parseMaterialNode(node);
+        }
+    }
+
+    ASSERT(has_geometry_node);
+
+    return visual_desc_;
+}
+
+UrdfCollisionDesc UrdfParser::parseCollisionNode(FXmlNode* collision_node)
+{
+    UrdfCollisionDesc collision_desc;
+
+    collision_desc.name_ = TCHAR_TO_UTF8(*collision_node->GetAttribute(TEXT("name")));
+
+    bool has_origin_node = false;
+    bool has_geometry_node = false;
+
+    for (auto& node : collision_node->GetChildrenNodes()) {
+        const FString& tag = node->GetTag();
+        if (tag.Equals(TEXT("origin"))) {
+            ASSERT(!has_origin_node);
+            has_origin_node = true;
+            collision_desc.origin_.SetLocation(parseVector(node->GetAttribute("xyz")));
+            collision_desc.origin_.SetRotation(parseRotation(node->GetAttribute("rpy")));
+        } else if (tag.Equals(TEXT("geometry"))) {
+            ASSERT(!has_geometry_node);
+            has_geometry_node = true;
+            collision_desc.geometry_ = parseGeometryNode(node);
+        }
+    }
+
+    ASSERT(has_geometry_node);
+
+    return collision_desc;
+}
+
+UrdfLinkDesc UrdfParser::parseLinkNode(FXmlNode* link_node)
+{
+    UrdfLinkDesc link_desc;
+    link_desc.name_ = TCHAR_TO_UTF8(*link_node->GetAttribute(TEXT("name")));
+
+    bool has_inertia_node = false;
+
+    for (auto& node : link_node->GetChildrenNodes()) {
+        FString tag = node->GetTag();
+        if (tag.Equals(TEXT("inertial"))) {
+            ASSERT(!has_inertia_node);
+            has_inertia_node = true;
+            link_desc.inertial_desc_ = parseInertialNode(node);
+        } else if (tag.Equals(TEXT("visual"))) {
+            UrdfVisualDesc visual_desc = parseVisualNode(node);
```

Comment:
```
            link_desc.visual_descs_.push_back(parseVisualNode(node));
```

Let's use `push_back` for consistency, and because there is no benefit to `emplace_back` in this case because this code effectively constructs a temporary object even when using `emplace_back`.

Reply:
fixed